### PR TITLE
Colorize Type Arguments and Punctuation

### PR DESCRIPTION
**Pull request by cloudRoutine**
some fixes for #2410

### CODE REVIEW

**Comment by saul** [comment]
I believe that the apostrophe before a type argument is part of the type name and so should be coloured the same. 

Also what's the rationale for having type arguments and types colourised differently?

**Comment by cloudRoutine** [comment]
@saul I commented the same thing on some other issue you opened a while back, it's not like that because I want it to be

**Comment by vasily-kirichenko** [comment]
image
image
image

**Comment by vasily-kirichenko** [comment]
Revert coloring well known types as keywords (and add `int8` and `uint8` into the list).

**Comment by cloudRoutine** [comment]
>  I think coloring constructs like `obj` and `int32` as keywords is a bad convention because these types can be aliased over but they're still colored as keywords and although `obj` is a class and `int32` is a struct they're colored the same as `module`,`begin`,and `end`

@vasily-kirichenko  is there a good reason to color them as keywords? If we're going to do things the way that C# does it then `<` & `.` should be colored like operators, but since they are delimiters I agree that they should be colored as punctuation.

File: src/fsharp/PrettyNaming.fs [hunk 1]
```diff
@@ -386,6 +386,19 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
             // by the call to String.forall; it is a fast check used to avoid the call if possible.
             || (s.[0] = '~' && String.forall (fun c -> c = '~') s)
 
+    let IsPunctuation s =
+        if System.String.IsNullOrEmpty s then false else
+        match s with
+        | "," | ";" | "|" | ":" | "." | "*"
```
Comment (vasily-kirichenko) [inline src/fsharp/PrettyNaming.fs]: I'm pretty sure `*` can be an operator _or_ punctuation. 

Check how punctuation is handled in `sformat.fs`, search for `tagPunctuation`.

File: src/fsharp/vs/service.fs [hunk 2]
```diff
@@ -1482,8 +1483,8 @@ type TypeCheckInfo
             | CNR(_, (Item.CustomBuilder _ | Item.CustomOperation _), ItemOccurence.Use, _, _, _, m) ->
                 Some (m, SemanticClassificationType.ComputationExpression)
             // well known type aliases get colored as keywords
-            | CNR(_, (Item.Types (n, _)), _, _, _, _, m) when keywordTypes.Contains(n) ->
-                Some (m, SemanticClassificationType.IntrinsicType)
```
Comment (vasily-kirichenko) [inline src/fsharp/vs/service.fs]: > is there a good reason to color them as keywords? 

They are intrinsic types. I'd like to add `option` and `result` to the list as well. And may be more.

**Comment by vasily-kirichenko** [comment]
> is there a good reason to color them as keywords? 

They are intrinsic types. I'd like to add `option` and `result` to the list as well. And may be more.

**Comment by cloudRoutine** [comment]
coloring them the same as keywords decreases the semantic information that the colorization provides. Plus it'll make any alias of those names look like an "intrinsic" type.

So then we color `obj` but not `System.Object`? What about `System.Type`? Where does it stop? Do we restrict ourselves to some set of namespaces?

It makes little sense, it adds no value, it shouldn't be done

**Comment by cloudRoutine** [comment]
@vasily-kirichenko The problem with what I just committed that I hadn't solved earlier is that

These chevrons get colored the same as the identifier they're next to


These chevrons still get colored as punctuation when they should be operators


The `'` and `^` in the type arguments are still getting classified as operators instead of as part of the identifier





As soon as its past the `'` `^` it's properly classified as an identifier

**Comment by vasily-kirichenko** [comment]
@cloudRoutine 

Because

1. Coloring some things as keywords is a matter of preference, you are the first person who don't like it :)
2. We don't have user settings dialog to customize it (and won't have for many months because we blocked by Roslyn, search for the related issue)

Please, revert the related part and open a separate PR to discuss it. Maybe you are right about types like `obj` and `int`, but I do want 

* `typeof`
* `typedefof`
* `sizeof`
* `nameof`
* `raise`
* `reraise`

to be highlighted as keywords. All of them are not functions, but looks like them.

I also think

* `ref`
* `failwith`
* `failwithf`

could be colored as keywords as well, because they are very common special purpose functions.

So I propose to:

* remove current well-known-types-as-keywords approach
* colorize as keywords the first group of intrinsic "functions" (`nameof`, etc)
* postpone colorizing `int`, `ref`, etc. until we have user setting dialog to customize it

**Comment by cloudRoutine** [comment]
These all seem reasonable to me
> - typeof
> - typedefof
> - sizeof
> - nameof
> - raise
> - reraise
> - ref

plus maybe `nativeptr` and `ilsigptr` since these and the above are all based on inline IL and could loosely be in the category of "stuff that could(should) only be defined in FSharp.Core

`failwith` and `failwithf` are simple functions anyone could write, I don't think they should get special treatment

**Comment by vasily-kirichenko** [comment]
@cloudRoutine wait, I'll make a PR with changes in keywords.

**Comment by vasily-kirichenko** [comment]
@cloudRoutine For some reason I cannot make a PR to your branch, merge this branch directly please https://github.com/vasily-kirichenko/visualfsharp/tree/colorization

**Comment by vasily-kirichenko** [comment]
@dotnet-bot test this please

**Comment by cloudRoutine** [comment]
the operators are now differentiated from signature delimiters 



the classification is still wrong




but I don't think this matters

**Comment by cartermp** [comment]
I think these are good changes.  Semantic colorization is a good thing to iterate on and listen to feedback for.  Clearly there are kinks to work out given the nature of type aliases in F#.  Good work!

**Comment by KevinRansom** [comment]
@cloudRoutine 
Hey mate ... do you want to take a look at the conflict.  I think I'm with you on not coring types as keywords :-)

Let's colour them as types for now until we can add an option to color them as keywords, for people who like that way.

Kevin

**Comment by cloudRoutine** [comment]
@KevinRansom I think the conflicts are resolved now

**Comment by KevinRansom** [comment]
@cloudRoutine 
````
Errors and Failures

1) Failed : Microsoft.VisualStudio.FSharp.Editor.Tests.Roslyn.HelpContextServiceTests.Namespaces.BeforeDot
  Expected: <Some(System)>
  But was:  null
at <StartupCode$VisualFSharp-Unittests>.$HelpContextServiceTests.TestF1Keywords@69.Invoke(FSharpOption`1 exp, FSharpOption`1 res) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\HelpContextServiceTests.fs:line 70
at Microsoft.FSharp.Collections.ListModule.loop@183-27[T1,T2](FSharpFunc`3 f, FSharpList`1 list1, FSharpList`1 list2) in D:\j\workspace\release_ci_pa---3f142ccc\src\fsharp\FSharp.Core\list.fs:line 186
at Microsoft.FSharp.Collections.ListModule.Iterate2[T1,T2](FSharpFunc`2 action, FSharpList`1 list1, FSharpList`1 list2) in D:\j\workspace\release_ci_pa---3f142ccc\src\fsharp\FSharp.Core\list.fs:line 189
at Microsoft.VisualStudio.FSharp.Editor.Tests.Roslyn.HelpContextServiceTests.Namespaces.BeforeDot() in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\HelpContextServiceTests.fs:line 145

2) Failed : Tests.LanguageService.General.UsingMSBuild.TokenInfo.TriggerClasses
Expected:
(Operator, Delimiter, MemberSelect)

but got:
(Punctuation, Delimiter, MemberSelect).
at UnitTests.TestLib.Salsa.AssertEqual[a](a expected, a actual) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\TestLib.Salsa.fs:line 23
at Tests.LanguageService.General.UsingMSBuild.TokenInfo.TriggerClasses() in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.General.fs:line 413

3) Failed : Tests.LanguageService.General.UsingProjectSystem.TokenInfo.TriggerClasses
Expected:
(Operator, Delimiter, MemberSelect)

but got:
(Punctuation, Delimiter, MemberSelect).
at UnitTests.TestLib.Salsa.AssertEqual[a](a expected, a actual) in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\TestLib.Salsa.fs:line 23
at Tests.LanguageService.General.UsingMSBuild.TokenInfo.TriggerClasses() in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\Tests.LanguageService.General.fs:line 413

Run Settings
    RuntimeFramework: V4.0
    RunAsX86: True
    WorkDirectory: D:\j\workspace\release_ci_pa---3f142ccc\release\net40\bin
    MaxAgents: 1
    NumberOfTestWorkers: 1
    Verbose: True


````

**Comment by cloudRoutine** [comment]
@KevinRansom all tests passing

**Pull request merged**